### ISSUE TITLE
Integrate `SparseEncoder` model from SentenceTransformers `v5`

### ISSUE THREAD

**Issue by arthurbr11** [embeddings-benchmark/mteb#2873]
Hello!

## Context
We released Sentence Transformers v5 yesterday, introducing a new `SparseEncoder` model category. These are a subclass of `SentenceTransformer` that output sparse vectors instead of dense embeddings.

## Feature Description
Add support for `SparseEncoder` models in MTEB. I tested the current version and found one main blocking issue in the `Sentence TransformerWrapper` to make it compatible : https://github.com/embeddings-benchmark/mteb/blob/f27648baec2086389b162b2b94026d2ed6325416/mteb/models/sentence_transformer_wrapper.py#L114

with the .numpy (applied on a sparse tensor it's breaking and not wanted anyway)

## Questions
1. Is there a specific reason for the `.numpy()` conversion? Can we make it conditional for sparse models?
2. For similarity computation, sparse models have an optimized `similarity` function that should be use this instead of MTEB's standard similarity functions, is it gonna be always the case ? Because if dense way of computing the similarity is used it will broke or be really long.
3. Since no sparse indexes exist in MTEB, encoding by chunk would be the best option to make sure the similarity compute is reasonable but is it plan to handle other indexes ? 

This would probably be part of a bigger refactor of the Sentence Transformers handling with https://github.com/embeddings-benchmark/mteb/issues/2871 and other possible feature.

Would love to hear your thoughts thist. 

cc @tomaarsen  

Arthur Bresnu

**Comment by Samoed** [comment]
1. What embedding type `SparseEncoder` expecting?
2. We've changed this in `v2` branch to use models similarity function if available, so I think this is resolved in `v2` https://github.com/embeddings-benchmark/mteb/blob/ed69e60c17177ab1c66358e99022e13ad4f7c184/mteb/models/abs_encoder.py#L35-L53
3. Yes, we have plans for new indexes https://github.com/embeddings-benchmark/mteb/issues/1593 https://github.com/embeddings-benchmark/mteb/issues/2728

**Comment by arthurbr11** [comment]
1. `SparseEncoder` models output a `torch.sparse_coo` by default.

2. That's great!

3. Nice! That means having one optimized for sparse vectors would be beneficial, it would be faster than the current model similarity computation.

**Comment by KennethEnevoldsen** [comment]
@arthurbr11 and @Samoed added a mock implementation of support for sparse embeddings (#2878). It seems like it will require a lot of adaptations, as the sparse torch matrix does not live up to Python's array API spec (see related issue here: #941).

@arthurbr11 would a solution be to convert the matrix to dense for cases such as classification, clustering etc., but keep it as a sparse vector for cases involving similarity measures?

I might be missing some important aspects (or smart tricks, so do let me know if there is anything I missed).

**Comment by arthurbr11** [comment]
You're right that we face compatibility challenges with sparse PyTorch matrices not conforming to Python's array API specification. To work around this, particularly for slicing operations, we've structured our encoder to return a list of sparse tensors instead of a single 2D sparse tensor when we want to do these operations. You can see our usage of this [here](https://github.com/UKPLab/sentence-transformers/blob/bb3886018177b3feab60f2fd3184fc2cd0cf8b54/sentence_transformers/sparse_encoder/evaluation/SparseRerankingEvaluator.py#L191) and [here](https://github.com/UKPLab/sentence-transformers/blob/bb3886018177b3feab60f2fd3184fc2cd0cf8b54/sentence_transformers/evaluation/RerankingEvaluator.py#L255) the reason why we did it .

Converting to dense would indeed be the simplest solution, but it's not practical at all. The memory would explode really quickly. We're working with tensors that can exceed 30K of dim with only hundreds of non zeros values.

For the task you mentioned, i'm not sure i really understand why you can't use the native sparse adapt similarity of the model and then do operation on the scores ? But i'm not really aware of how these tasks evaluation work and i'm probably missing some difficulties.

**Comment by KennethEnevoldsen** [comment]
Ahh, that's great to hear - I think that could potentially solve the issues. We would have to expand our notion of an array a bit, but I think it might work.

> Converting to dense would indeed be the simplest solution, but it's not practical at all. The memory would explode really quickly. We're working with tensors that can exceed 30K of dim with only hundreds of non zeros values.

> For the task you mentioned, i'm not sure i really understand why you can't use the native sparse adapt similarity of the model and then do operation on the scores ? But i'm not really aware of how these tasks evaluation work and i'm probably missing some difficulties.

Ahh the classification and clustering etc, rely on the sklearn interface which requires a dense vector (or a sparse numpy vector)

**Comment by arthurbr11** [comment]
I think you were refering to sparse scipy vectors as there isn't numpy ones. If they are handled, then I think we can easily have this format from the torch sparse ones.

**Comment by KennethEnevoldsen** [comment]
Perfect, that seems to be the best way then - this is currently not first on my to-do list though (however, it is good to have an overview of what the problem entails). First priority is v2, and then we can add this in a backward-compatible manner.

**Comment by thibault-formal** [comment]
hey @arthurbr11 @KennethEnevoldsen 

Thanks for looking into this! I’m very interested in this integration—we’re developing new sparse models that we’d like to evaluate on MTEB as well (especially for tasks beyond retrieval), so it would definitely be a cool feature.

I’ve started experimenting a bit with MTEB and our custom model. Retrieval evaluation works almost out of the box as long as the model has the right `encode` method. I initially thought it would be more complicated, since I assumed MTEB was indexing everything under the hood, but by default it seems to just keep the top-k documents for each query (if I understood correctly).

I’m still getting familiar with MTEB’s logic, but it looks like the main challenges are around handling other tasks, like classification, with sparse embeddings? For example, in your mock implementation @KennethEnevoldsen, there’s no modification on the retrieval side.

Do you have an estimate of when this integration might be available? I’d be happy to help if needed.

**Comment by KennethEnevoldsen** [comment]
> I initially thought it would be more complicated, since I assumed MTEB was indexing everything under the hood, but by default it seems to just keep the top-k documents for each query (if I understood correctly).

v2 adds an index step, so that might help you  (generally adds a clear search interface). This should, in general, support a broader set of search systems

> Do you have an estimate of when this integration might be available? I’d be happy to help if needed.

We will probably include it after v2 is done (which is getting close)

**Comment by thibault-formal** [comment]
ok great! 
I will keep an eye on this issue then!
